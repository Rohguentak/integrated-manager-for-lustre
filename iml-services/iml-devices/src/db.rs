// Copyright (c) 2019 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{
    change::{self, Change},
    error::ImlDevicesError,
};
use futures::TryStreamExt;
use iml_postgres::{select_all, Client, Transaction};
use iml_wire_types::{
    db::{Device, DeviceHost, DeviceId, DeviceIds, DeviceType, MountPath, Name, Paths, Size},
    Fqdn,
};
use std::{
    collections::{BTreeMap, BTreeSet},
    iter,
    path::PathBuf,
};

pub type FlatDevices = BTreeMap<DeviceId, FlatDevice>;

pub type Devices = BTreeMap<DeviceId, Device>;

pub type DevicesRef<'a> = BTreeMap<&'a DeviceId, &'a Device>;

pub type DeviceHostKey = (DeviceId, Fqdn);

pub type DeviceHosts = BTreeMap<DeviceHostKey, DeviceHost>;

pub type DeviceHostsRef<'a> = BTreeMap<&'a DeviceHostKey, &'a DeviceHost>;

/// A device (Block or Virtual).
/// These should be unique per cluster
#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct FlatDevice {
    pub id: DeviceId,
    pub size: u64,
    pub device_type: DeviceType,
    pub parents: BTreeSet<DeviceId>,
    // pub usable_for_lustre: bool,
    pub children: BTreeSet<DeviceId>,
    pub paths: BTreeSet<PathBuf>,
    pub mount_path: Option<PathBuf>,
    pub fs_type: Option<String>,
    pub fs_label: Option<String>,
    pub fs_uuid: Option<String>,
}

fn create_dev(
    flat_dev: &FlatDevice,
    fqdn: Fqdn,
) -> ((DeviceId, Device), (DeviceHostKey, DeviceHost)) {
    let d = Device {
        id: flat_dev.id.clone(),
        size: Size(flat_dev.size),
        device_type: flat_dev.device_type.clone(),
        parents: DeviceIds(flat_dev.parents.clone()),
        children: DeviceIds(flat_dev.children.clone()),
        // usable_for_lustre: flat_dev.usable_for_lustre,
        usable_for_lustre: false,
    };

    let dh = DeviceHost {
        device_id: flat_dev.id.clone(),
        fqdn,
        local: true,
        paths: Paths(flat_dev.paths.clone()),
        mount_path: MountPath(flat_dev.mount_path.clone()),
        fs_type: flat_dev.fs_type.clone(),
        fs_label: flat_dev.fs_label.clone(),
        fs_uuid: flat_dev.fs_uuid.clone(),
    };

    (
        (flat_dev.id.clone(), d),
        ((flat_dev.id.clone(), dh.fqdn.clone()), dh),
    )
}

pub fn convert_flat_devices(flat_devices: &FlatDevices, fqdn: Fqdn) -> (Devices, DeviceHosts) {
    flat_devices
        .iter()
        .map(|x| create_dev(x.1, fqdn.clone()))
        .unzip()
}

fn get_fqdns<'a>(device_hosts: &'a DeviceHosts) -> impl Iterator<Item = &'a Fqdn> {
    device_hosts.iter().map(|((d, fqdn), v)| fqdn)
}

/// Given a device id and some `DeviceHosts`,
/// filter to the cooresponding hosts.
fn filter_device_hosts<'a>(
    id: &'a DeviceId,
    device_hosts: &'a DeviceHosts,
) -> impl Iterator<Item = (&'a DeviceHostKey, &'a DeviceHost)> {
    device_hosts.iter().filter(move |(_, v)| &v.device_id == id)
}

/// Given a device id and some `DeviceHosts`,
/// try to find the first cooresponding host.
fn find_device_host<'a>(
    id: &'a DeviceId,
    device_hosts: &'a DeviceHostsRef<'a>,
) -> Option<(&'a &'a DeviceHostKey, &'a &'a DeviceHost)> {
    device_hosts
        .into_iter()
        .find(move |(_, v)| &v.device_id == id)
}

pub fn get_local_device_hosts<'a>(
    device_hosts: &'a DeviceHosts,
    fqdn: &Fqdn,
) -> DeviceHostsRef<'a> {
    device_hosts
        .into_iter()
        .filter(|(_k, v)| &v.fqdn == fqdn && v.local)
        .collect()
}

pub fn get_local_devices<'a>(
    local_device_hosts: &DeviceHostsRef<'_>,
    devices: &'a Devices,
) -> DevicesRef<'a> {
    devices
        .into_iter()
        .filter(|(k, _)| {
            local_device_hosts
                .into_iter()
                .find(|(_, v)| &&v.device_id == k)
                .is_some()
        })
        .collect()
}

pub fn get_other_device_hosts<'a>(
    db_device_hosts: &'a DeviceHosts,
    fqdn: &'a Fqdn,
) -> DeviceHostsRef<'a> {
    db_device_hosts
        .iter()
        .filter(move |(_, v)| &v.fqdn != fqdn)
        .collect()
}

pub fn get_devices_by_device_host<'a>(
    device_hosts: &'a DeviceHostsRef<'a>,
    devices: &'a Devices,
) -> DevicesRef<'a> {
    devices
        .iter()
        .filter(move |(k, _)| find_device_host(&k, device_hosts).is_some())
        .collect()
}

pub async fn get_db_devices(mut client: &mut Client) -> Result<Devices, iml_postgres::Error> {
    select_all(
        &mut client,
        &format!("SELECT * FROM {}", Device::table_name()),
        iter::empty(),
    )
    .await?
    .map_ok(Device::from)
    .map_ok(|x| (x.id.clone(), x))
    .try_collect()
    .await
}

pub async fn get_db_device_hosts(
    mut client: &mut Client,
) -> Result<Vec<DeviceHost>, iml_postgres::Error> {
    select_all(
        &mut client,
        &format!("SELECT * FROM {}", DeviceHost::table_name()),
        iter::empty(),
    )
    .await?
    .map_ok(DeviceHost::from)
    .try_collect()
    .await
}

async fn insert_device_host(
    transaction: &mut Transaction<'_>,
    fqdn: &Fqdn,
    x: &DeviceHost,
) -> Result<(), ImlDevicesError> {
    let s = transaction.prepare(
        &format!("INSERT INTO {} (device_id, fqdn, local, paths, mount_path, fs_type, fs_label, fs_uuid) VALUES ($1, $2, $3, $4, $5, $6, $7, $8)", DeviceHost::table_name())
    ).await?;

    transaction
        .execute(
            &s,
            &[
                &x.device_id,
                &fqdn.0,
                &x.local,
                &x.paths,
                &x.mount_path,
                &x.fs_type,
                &x.fs_label,
                &x.fs_uuid,
            ],
        )
        .await?;

    Ok(())
}

async fn update_device_host(
    transaction: &mut Transaction<'_>,
    fqdn: &Fqdn,
    x: &DeviceHost,
) -> Result<(), ImlDevicesError> {
    let s = transaction.prepare(&format!("UPDATE {} SET local = $3, paths = $4, mount_path = $5, fs_type = $6, fs_label= $7, fs_uuid=$8 WHERE device_id = $1 AND fqdn = $2", DeviceHost::table_name())).await?;

    transaction
        .execute(
            &s,
            &[
                &x.device_id,
                &fqdn.0,
                &x.local,
                &x.paths,
                &x.mount_path,
                &x.fs_type,
                &x.fs_label,
                &x.fs_uuid,
            ],
        )
        .await?;

    Ok(())
}

async fn remove_device_host(
    transaction: &mut Transaction<'_>,
    fqdn: &Fqdn,
    id: &DeviceId,
) -> Result<(), ImlDevicesError> {
    let s = transaction
        .prepare(&format!(
            "DELETE FROM {} WHERE device_id = $1 AND fqdn = $2",
            DeviceHost::table_name()
        ))
        .await?;

    transaction.execute(&s, &[id, &fqdn.0]).await?;

    Ok(())
}

pub async fn persist_local_device_hosts<'a>(
    mut transaction: &mut Transaction<'a>,
    incoming_devices: &DeviceHosts,
    local_db_device_hosts: &DeviceHostsRef<'_>,
) -> Result<(), ImlDevicesError> {
    for c in change::get_changes_values(local_db_device_hosts, &incoming_devices.iter().collect()) {
        match c {
            Change::Add(d) => {
                tracing::debug!(
                    "Going to insert new devicehost {:?}, {:?}",
                    d.fqdn,
                    d.device_id
                );

                insert_device_host(&mut transaction, &d.fqdn, d).await?;
            }
            Change::Update(d) => {
                tracing::debug!("Going to update devicehost {:?}, {:?}", d.fqdn, d.device_id);

                update_device_host(&mut transaction, &d.fqdn, d).await?;
            }
            Change::Remove(d) => {
                tracing::debug!("Going to remove devicehost {:?}, {:?}", d.fqdn, d.device_id);

                remove_device_host(&mut transaction, &d.fqdn, &d.device_id).await?;
            }
        };
    }

    Ok(())
}

pub async fn persist_local_devices<'a>(
    transaction: &mut Transaction<'a>,
    incoming_devices: &Devices,
    other_devices: &DevicesRef<'a>,
    local_db_devices: &DevicesRef<'a>,
) -> Result<(), ImlDevicesError> {
    for c in change::get_changes_values(&local_db_devices, &incoming_devices.iter().collect()) {
        match c {
            Change::Add(d) => {
                if other_devices.get(&d.id).is_some() {
                    tracing::info!("Device {:?} already added by another host.", d.id);
                    continue;
                }

                tracing::debug!("Going to add device {:?}", d.id);

                let s = transaction.prepare("INSERT INTO chroma_core_device (id, size, usable_for_lustre, device_type, parents, children) VALUES ($1, $2, $3, $4, $5, $6)").await?;

                transaction
                    .execute(
                        &s,
                        &[
                            &d.id,
                            &d.size,
                            &d.usable_for_lustre,
                            &d.device_type,
                            &d.parents,
                            &d.children,
                        ],
                    )
                    .await?;
            }
            Change::Update(d) => {
                tracing::debug!("Going to update device {:?}", d.id);

                let s = transaction.prepare("UPDATE chroma_core_device SET size = $2, usable_for_lustre = $3, device_type = $4, parents=$5, children=$6 WHERE id = $1").await?;

                transaction
                    .execute(
                        &s,
                        &[
                            &d.id,
                            &d.size,
                            &d.usable_for_lustre,
                            &d.device_type,
                            &d.parents,
                            &d.children,
                        ],
                    )
                    .await?;
            }
            Change::Remove(d) => {
                // @TODO: I think devices should probably not be deleted.
                // Orphan devices should probably be surfaced as alerts.
                tracing::debug!("Going to remove device {:?}", d.id);

                let s = transaction
                    .prepare(&format!(
                        "DELETE FROM {} WHERE id = $1",
                        Device::table_name()
                    ))
                    .await?;

                transaction.execute(&s, &[&d.id]).await?;
            }
        }
    }

    Ok(())
}

/// Some devices should appear on multiple hosts even if they are physically existent on one host.
///
/// Examples are Zpools / Datasets, LVs / VGs and MdRaid.
pub async fn update_virtual_devices<'a>(
    transaction: &mut Transaction<'a>,
    fqdn: &Fqdn,
    incoming_devices: &Devices,
    db_device_hosts: &DeviceHosts,
    flat_devices: &FlatDevices,
) -> Result<(), ImlDevicesError> {
    for virtual_device in incoming_devices.values() {
        tracing::info!("virtual_device: {:#?}", virtual_device);
        let virtual_device_flat = flat_devices.get(&virtual_device.id);
        tracing::info!("virtual_device_flat: {:#?}", virtual_device_flat);

        let mut parents = virtual_device.parents.clone();

        let mut depth = 1;
        let max_depth = 8;

        let mut transaction_device_hosts = BTreeMap::new();
        while depth < max_depth {
            tracing::info!("depth = {}, parents = {:#?}", depth, parents);
            let mut new_parents = BTreeSet::new();

            for parent in parents.iter() {
                let other_hosts: Vec<_> = filter_device_hosts(&parent, &db_device_hosts)
                    .filter(|(_, v)| &v.fqdn != fqdn)
                    .map(|(_, v)| v)
                    .collect();

                for other_host in other_hosts {
                    let other_device_host = DeviceHost {
                        device_id: virtual_device.id.clone(),
                        fqdn: other_host.fqdn.clone(),
                        local: true,
                        // Does it make sense to import paths from other hosts?
                        paths: Paths(
                            virtual_device_flat
                                .map(|x| x.paths.clone())
                                .unwrap_or(BTreeSet::new()),
                        ),
                        // It can't be mounted on other hosts at the time this is processed?
                        mount_path: MountPath(None),
                        fs_type: virtual_device_flat.map(|x| x.fs_type.clone()).unwrap_or(None),
                        fs_label: virtual_device_flat.map(|x| x.fs_label.clone()).unwrap_or(None),
                        fs_uuid: virtual_device_flat.map(|x| x.fs_uuid.clone()).unwrap_or(None),
                    };

                    if db_device_hosts
                        .get(&(virtual_device.id.clone(), other_host.fqdn.clone()))
                        .is_none()
                        && transaction_device_hosts
                            .get(&(virtual_device.id.clone(), other_host.fqdn.clone()))
                            .is_none()
                    {
                        // TODO: Produce changes instead of directly mutating the transaction
                        insert_device_host(transaction, &other_host.fqdn, &other_device_host)
                            .await
                            .unwrap();
                        transaction_device_hosts.insert(
                            (virtual_device.id.clone(), other_host.fqdn.clone()),
                            other_device_host,
                        );
                    } else {
                        update_device_host(transaction, &other_host.fqdn, &other_device_host)
                            .await
                            .unwrap();
                    }
                }

                for (id, db_host) in db_device_hosts {
                    let &(ref device_id, _) = id;
                    let device = flat_devices.get(device_id);
                    if let Some(d) = device {
                        let parents = &d.parents;
                        for parent in parents {
                            if db_device_hosts
                                .get(&(parent.clone(), db_host.fqdn.clone()))
                                .is_none()
                                && transaction_device_hosts
                                    .get(&(parent.clone(), db_host.fqdn.clone()))
                                    .is_none()
                            {
                                remove_device_host(
                                    transaction,
                                    &db_host.fqdn.clone(),
                                    &virtual_device.id.clone(),
                                )
                                .await
                                .unwrap();
                            }
                        }
                    }
                }

                flat_devices.get(parent).map(|x| {
                    for p in &x.parents {
                        new_parents.insert(p.clone());
                    }
                });
            }

            if new_parents.is_empty() {
                break;
            }
            parents = new_parents;
            depth += 1;
        }

        if depth == max_depth {
            tracing::error!("Hit upper limit {} on recursion", max_depth);
        }
    }

    Ok(())
}

// pub async fn perform_updates<'a>(
//     mut transaction: &mut Transaction<'a>,
//     fqdn: Fqdn,
//     dhs: DeviceHosts,
//     devices: Devices,
//     db_dhs: DeviceHosts,
//     db_devices: Devices,
// ) -> Result<(), ImlDevicesError> {
//     let local_dhs = get_local_device_hosts(&db_dhs, &fqdn);

//     let other_hosts: Vec<_> = db_dhs
//         .values()
//         .filter(|x| x.fqdn != fqdn)
//         .filter(|x| !x.local)
//         .collect();

//     let dhs_changes = change::get_changes(&local_dhs.iter().collect(), &dhs.iter().collect())
//         .into_iter()
//         .collect::<Vec<_>>();

//     let remote_dhs_changes: Vec<Change<DeviceHostKey>> = other_hosts
//         .iter()
//         .filter(|x| {
//             local_dhs
//                 .iter()
//                 .find(|(_, v)| x.device_id == v.device_id)
//                 .is_some()
//         })
//         .fold(vec![], |mut acc, x| {
//             let change = dhs_changes.iter().find(|&y| y.0 == x.device_id);

//             if let Some(change) = change {
//                 match change {
//                     Change::Add(v) => tracing::warn!(
//                         "Unexepected Add when comparing {:?} (remote) with {:?} (local)",
//                         x,
//                         v
//                     ),
//                     Change::Update(v) => acc.push(Change::Update((v.0.clone(), x.fqdn.clone()))),
//                     Change::Remove(v) => acc.push(Change::Remove((v.0.clone(), x.fqdn.clone()))),
//                 }
//             }

//             acc
//         });

//     let device_changes =
//         change::get_changes(&db_devices.iter().collect(), &devices.iter().collect());

//     // 3. Find any new non-local devices and ensure they are present on all relevant nodes.
//     let remote_dhs_changes = device_changes
//         .iter()
//         .filter(|x| x.is_add())
//         .filter_map(|x| devices.get(x))
//         .fold(remote_dhs_changes, |mut acc, x| {
//             for (fqdn, ys) in db_dhs.iter() {
//                 if x.parents
//                     .iter()
//                     .collect::<BTreeSet<&DeviceId>>()
//                     .is_subset(&ys.keys().collect())
//                 {
//                     acc.push(Change::Add((fqdn.clone(), x.id.clone())));
//                 }
//             }

//             acc
//         });

//     // 4. Insert changes into the DB
//     for c in remote_dhs_changes {
//         match c {
//             Change::Add((fqdn, id)) => {
//                 tracing::info!("Going to insert new remote devicehost {:?}, {:?}", fqdn, id);

//                 let x = dhs.get(&id).unwrap();

//                 insert_device_host(&mut transaction, &fqdn, x).await?;
//             }
//             Change::Update((fqdn, id)) => {
//                 tracing::info!("Going to update remote devicehost {:?}, {:?}", fqdn, id);

//                 let x = dhs.get(&id).unwrap();

//                 update_device_host(&mut transaction, &fqdn, x).await?;
//             }
//             Change::Remove((fqdn, id)) => {
//                 tracing::info!("Going to delete remote devicehost {:?}, {:?}", fqdn, id);

//                 remove_device_host(&mut transaction, &fqdn, &id).await?;
//             }
//         }
//     }

//     Ok(())
// }
