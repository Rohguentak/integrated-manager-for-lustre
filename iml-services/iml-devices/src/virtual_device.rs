use crate::{
    breadth_first_parent_iterator::BreadthFirstParentIterator,
    change::Change,
    db::{DeviceHosts, Devices},
    error::ImlDevicesError,
};
use iml_wire_types::db::{Device, DeviceHost, DeviceId, DeviceType, MountPath, Paths};
use iml_wire_types::Fqdn;
use std::collections::{BTreeMap, BTreeSet};

fn is_virtual_device(device: &Device) -> bool {
    device.device_type == DeviceType::MdRaid
        || device.device_type == DeviceType::VolumeGroup
        || device.device_type == DeviceType::LogicalVolume
        || device.device_type == DeviceType::Dataset
        || device.device_type == DeviceType::Zpool
}

fn make_other_device_host(
    device_id: DeviceId,
    fqdn: Fqdn,
    virtual_device_host: Option<&DeviceHost>,
) -> DeviceHost {
    DeviceHost {
        device_id,
        fqdn,
        local: true,
        // Does it make sense to use paths from other hosts?
        paths: Paths(
            virtual_device_host
                .map(|x| x.paths.clone())
                .unwrap_or(BTreeSet::new()),
        ),
        // It can't be mounted on other hosts at the time this is processed?
        mount_path: MountPath(None),
        fs_type: virtual_device_host
            .map(|x| x.fs_type.clone())
            .unwrap_or(None),
        fs_label: virtual_device_host
            .map(|x| x.fs_label.clone())
            .unwrap_or(None),
        fs_uuid: virtual_device_host
            .map(|x| x.fs_uuid.clone())
            .unwrap_or(None),
    }
}

fn make_other_device_host_for_removal(device_id: DeviceId, fqdn: Fqdn) -> DeviceHost {
    DeviceHost {
        device_id,
        fqdn,
        local: true,
        paths: Paths(BTreeSet::new()),
        mount_path: MountPath(None),
        fs_type: None,
        fs_label: None,
        fs_uuid: None,
    }
}

fn add_device_host(
    device_id: DeviceId,
    fqdn: Fqdn,
    virtual_device_host: Option<&DeviceHost>,
    results: &mut BTreeMap<(DeviceId, Fqdn), Change<DeviceHost>>,
) {
    let other_device_host =
        make_other_device_host(device_id.clone(), fqdn.clone(), virtual_device_host);

    tracing::info!(
        "Adding new device host with id {:?} to host {:?}",
        device_id,
        fqdn
    );

    results.insert(
        (device_id.clone(), fqdn.clone()),
        Change::Add(other_device_host),
    );
}

fn update_device_host(
    device_id: DeviceId,
    fqdn: Fqdn,
    virtual_device_host: Option<&DeviceHost>,
    results: &mut BTreeMap<(DeviceId, Fqdn), Change<DeviceHost>>,
) {
    let other_device_host =
        make_other_device_host(device_id.clone(), fqdn.clone(), virtual_device_host);

    tracing::info!(
        "Updating device host with id {:?} on host {:?}",
        device_id,
        fqdn
    );

    results.insert(
        (device_id.clone(), fqdn.clone()),
        Change::Update(other_device_host),
    );
}

fn remove_device_host(
    device_id: DeviceId,
    fqdn: Fqdn,
    results: &mut BTreeMap<(DeviceId, Fqdn), Change<DeviceHost>>,
) {
    let other_device_host = make_other_device_host_for_removal(device_id.clone(), fqdn.clone());

    tracing::info!(
        "Removing device host with id {:?} on host {:?}",
        device_id,
        fqdn,
    );
    results.insert(
        (device_id.clone(), fqdn.clone()),
        Change::Remove(other_device_host),
    );
}

fn are_all_parents_available(
    devices: &Devices,
    device_hosts: &DeviceHosts,
    host: Fqdn,
    child_id: &DeviceId,
) -> bool {
    let mut i = BreadthFirstParentIterator::new(devices, child_id);
    let all_available = i.all(|p| {
        let result = device_hosts.get(&(p.clone(), host.clone())).is_some();
        tracing::info!("Checking device {:?} on host {:?}: {:?}", p, host, result);
        result
    });
    tracing::info!(
        "Host: {:?}, device: {:?}, all_available: {:?}",
        host,
        child_id,
        all_available
    );
    all_available
}

fn are_all_parents_available_with_results(
    devices: &Devices,
    device_hosts: &DeviceHosts,
    results: &BTreeMap<(DeviceId, Fqdn), Change<DeviceHost>>,
    host: Fqdn,
    child_id: &DeviceId,
) -> bool {
    let mut i = BreadthFirstParentIterator::new(devices, child_id);
    let all_available = i.all(|p| {
        let result = device_hosts.get(&(p.clone(), host.clone())).is_some();
        tracing::info!("Checking device {:?} on host {:?}: {:?}", p, host, result);
        let result_results = if !result {
            let result_results = results.get(&(p.clone(), host.clone())).is_some();
            tracing::info!(
                "Checking device {:?} on host {:?} in results: {:?}",
                p,
                host,
                result_results
            );
            Some(result_results)
        } else {
            None
        };

        result || result_results.unwrap()
    });
    tracing::info!(
        "Host: {:?}, device: {:?}, all_available: {:?}",
        host,
        child_id,
        all_available
    );
    all_available
}

pub fn compute_virtual_device_changes<'a>(
    fqdn: &Fqdn,
    incoming_devices: &Devices,
    incoming_device_hosts: &DeviceHosts,
    db_devices: &Devices,
    db_device_hosts: &DeviceHosts,
) -> Result<BTreeMap<(DeviceId, Fqdn), Change<DeviceHost>>, ImlDevicesError> {
    tracing::info!(
        "Incoming: devices: {}, device hosts: {}, Database: devices: {}, device hosts: {}",
        incoming_devices.len(),
        incoming_device_hosts.len(),
        db_devices.len(),
        db_device_hosts.len()
    );
    let mut results = BTreeMap::new();

    let virtual_devices = incoming_devices
        .iter()
        .filter(|(_, d)| is_virtual_device(d))
        .map(|(_, d)| d);
    let vd2 = virtual_devices.clone();
    let vd3 = virtual_devices.clone();

    // We're iterating the device twice
    // Consider devices h -> g -> f -> e
    // f and e are virtual
    // e will be seen first and it won't have f as a parent available on the other host
    // so it won't be added
    // f will be added
    // then on second iteration e will find previously added f and will be added

    // Limitation: we support up to three levels currently
    // We need to look into more generic algorithm in the future
    for virtual_device in virtual_devices.chain(vd2).chain(vd3) {
        tracing::info!(
            "virtual_device: {:?}, parents: {:?}, children: {:?}",
            virtual_device.id,
            virtual_device.parents,
            virtual_device.children
        );
        let virtual_device_host =
            incoming_device_hosts.get(&(virtual_device.id.clone(), fqdn.clone()));
        tracing::info!(
            "virtual_device_host: {:?}, fqdn: {:?}",
            virtual_device_host.map(|x| &x.device_id),
            virtual_device_host.map(|x| &x.fqdn)
        );

        // For this host itself, run parents check for this virtual device ON THE INCOMING DEVICE HOSTS
        // If it fails, remove the device host of this virtual device FROM THE DB
        // We don't add virtual devices here because either:
        // 1. Device-scanner sends us the virtual device, if it's initially available on this host
        // 2. We will add it while processing other hosts, if it's not initially available on this host
        {
            let all_available = are_all_parents_available(
                incoming_devices,
                incoming_device_hosts,
                fqdn.clone(),
                &virtual_device.id,
            );
            if !all_available {
                // remove from db if present
                let is_in_db = db_device_hosts
                    .get(&(virtual_device.id.clone(), fqdn.clone()))
                    .is_some();

                if is_in_db {
                    remove_device_host(virtual_device.id.clone(), fqdn.clone(), &mut results);
                }
            }
        }

        // For all other hosts, run parents check for this virtual device ON THE DB.
        // This is because main loop processes updates from single host at a time.
        // That means current state of other hosts is in DB at this point.

        let all_other_host_fqdns: BTreeSet<_> = db_device_hosts
            .iter()
            .filter_map(|((_, f), _)| if f != fqdn { Some(f) } else { None })
            .collect();

        for host in all_other_host_fqdns {
            // We want to look at in-flight changes (results)?
            let all_available = are_all_parents_available_with_results(
                incoming_devices,
                &db_device_hosts,
                &results,
                host.clone(),
                &virtual_device.id,
            );

            let is_in_db = db_device_hosts
                .get(&(virtual_device.id.clone(), host.clone()))
                .is_some();

            if all_available {
                // add to database if missing and not in flight
                // update in database if present and not in flight
                // update in flight if in flight - is it necessary though?
                let is_in_results = results
                    .get(&(virtual_device.id.clone(), host.clone()))
                    .is_none();

                if !is_in_db && is_in_results {
                    add_device_host(
                        virtual_device.id.clone(),
                        host.clone(),
                        virtual_device_host,
                        &mut results,
                    );
                } else if is_in_db && is_in_results {
                    update_device_host(
                        virtual_device.id.clone(),
                        host.clone(),
                        virtual_device_host,
                        &mut results,
                    );
                } else {
                    let is_in_incoming = incoming_device_hosts
                        .get(&(virtual_device.id.clone(), fqdn.clone()))
                        .is_some();

                    tracing::warn!(
                        "DB: {:?}, incoming: {:?}, results: {:?}",
                        is_in_db,
                        is_in_incoming,
                        is_in_results
                    );
                }
            } else {
                // remove from db if present and not in flight
                // remove from in-flight if in flight - is it necessary though?

                if is_in_db {
                    remove_device_host(virtual_device.id.clone(), host.clone(), &mut results);
                } else {
                    // It wasn't in the DB in the first place, nothing to do
                }
            }
        }
    }

    Ok(results)
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::db::test::{deser_devices, deser_fixture};
    use ::test_case::test_case;
    use insta::assert_debug_snapshot;

    #[test_case("vd_with_shared_parents_added_to_oss2")]
    #[test_case("vd_with_no_shared_parents_not_added_to_oss2")]
    // A leaf device has changed data on one host
    // It has to have updated data on the other one
    #[test_case("vd_with_shared_parents_updated_on_oss2")]
    // A leaf device is replaced with another leaf device
    // Previous one stays in the DB as available
    // We're not removing it since parents are still available
    #[test_case("vd_with_shared_parents_replaced_on_oss2")]
    #[test_case("vd_with_shared_parents_removed_from_oss2_when_parent_disappears")]
    #[test_case("vd_with_two_levels_of_shared_parents_added_to_oss2")]
    // An intermediary non-virtual parent disappears from the host
    // Its children are getting removed
    // Virtual devices from the other host receive updates that aren't necessary but aren't harmful
    #[test_case("vd_with_two_levels_of_shared_parents_removed_from_oss2_when_parent_disappears")]
    // A leaf virtual device is replaced with another virtual device on one host
    // Previous one stays in the DB as available
    // We're not removing it since parents are still available
    // Virtual device that is parent of the replaced receives update that isn't necessary but isn't harmful
    #[test_case("vd_with_two_levels_of_shared_parents_replaced_on_oss2")]
    // A leaf device has changed data on one host
    // It has to have updated data on the other one
    // Virtual device that is parent of the updated receives update that isn't necessary but isn't harmful
    #[test_case("vd_with_two_levels_of_shared_parents_updated_on_oss2")]
    #[test_case("vd_with_two_levels_of_shared_parents_in_reverse_order_added_to_oss2")]
    #[test_case("vd_with_three_levels_of_shared_parents_in_reverse_order_added_to_oss2")]
    fn compute_virtual_device_changes(test_name: &str) {
        crate::db::test::_init_subscriber();
        let (fqdn, incoming_devices, incoming_device_hosts, db_devices, db_device_hosts) =
            deser_fixture(test_name);

        let updates = super::compute_virtual_device_changes(
            &fqdn,
            &incoming_devices,
            &incoming_device_hosts,
            &db_devices,
            &db_device_hosts,
        )
        .unwrap();

        assert_debug_snapshot!(test_name, updates);
    }

    #[test_case("single_device_doesnt_iterate", "a")]
    #[test_case("single_parent_produces_one_item", "b")]
    #[test_case("two_parents_produce_two_items", "c")]
    #[test_case("parent_and_double_parent_produce_three_items", "c1")]
    #[test_case("triple_parent_and_double_parent_produce_five_items", "c1")]
    fn breadth_first_iterator(test_case: &str, child: &str) {
        let prefix = String::from("fixtures/") + test_case + "/";
        let devices = deser_devices(prefix.clone() + "devices.json");

        let id = DeviceId(child.into());
        let i = BreadthFirstParentIterator::new(&devices, &id);
        let result: Vec<_> = i.collect();
        assert_debug_snapshot!(test_case, result);
    }
}
