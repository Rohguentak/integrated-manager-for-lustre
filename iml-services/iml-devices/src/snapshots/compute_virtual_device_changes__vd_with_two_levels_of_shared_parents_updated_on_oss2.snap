---
source: iml-services/iml-devices/src/virtual_device.rs
expression: updates
---
{
    (
        DeviceId(
            "c",
        ),
        Fqdn(
            "oss2",
        ),
    ): Update(
        DeviceHost {
            device_id: DeviceId(
                "c",
            ),
            fqdn: Fqdn(
                "oss2",
            ),
            local: true,
            paths: Paths(
                {
                    "/fake/path/zpool/a",
                },
            ),
            mount_path: MountPath(
                None,
            ),
            fs_type: Some(
                "some_fs",
            ),
            fs_label: Some(
                "some_label",
            ),
            fs_uuid: Some(
                "some_uuid",
            ),
        },
    ),
    (
        DeviceId(
            "d",
        ),
        Fqdn(
            "oss2",
        ),
    ): Update(
        DeviceHost {
            device_id: DeviceId(
                "d",
            ),
            fqdn: Fqdn(
                "oss2",
            ),
            local: true,
            paths: Paths(
                {
                    "/fake/path/dataset/a",
                },
            ),
            mount_path: MountPath(
                None,
            ),
            fs_type: Some(
                "some_fs",
            ),
            fs_label: Some(
                "new_some_label",
            ),
            fs_uuid: Some(
                "some_uuid",
            ),
        },
    ),
}
